src/test/scala/calc/RunTest.scala

```diff
@@ -38,4 +39,48 @@ class RunTest {
     assertRun(54.3, "54.3")
   }
 
+  @Test def runBinaryOp(): Unit = {
+    for (i <- 1 to 5){
+      val a = r.nextDouble()
+      val b = r.nextDouble()
+      assertRun(a + b, s"${a} + ${b}")
+      assertRun(a - b, s"${a} - ${b}")
+      assertRun(a * b, s"${a} * ${b}")
+      assertRun(a / b, s"${a} / ${b}")
+    }
+  }
+
+  @Test def runLetOp(): Unit = {
+    assertRun(123.321 + 456.654, "let x=123.321 in x+456.654")
+    assertRun(1.5 + 2.0, "let x=1.5 in let y=2.0 in x+y")
+  }
+
+  @Test def runCondOp(): Unit = {
+    assertRun(4.2, "if (0) 1.0 else 4.2")
+    assertRun(2.4, "if (1) 2.4 else 0.0")
+  }
+
+  @Test def runFindCapture(): Unit = {
+    def testCapture(p: String, res: Set[String]): Unit ={
+      val tree = Parser.parse(p).get.value
+      assertEquals(Compiler findCapture tree, res)
+    }
+
+    testCapture("let x=1 in x+y", Set("y"))
+    testCapture("let x=1 in let y=2 in u+v", Set("u","v"))
+    testCapture("let f=fun (x) = {let y=1 in x+y} in f (1)", Set())
+    testCapture("let f=fun (g) = {g (1)} in let g=fun (x) = {x+1} in f (g)", Set())
+    testCapture("let f=fun (g) = {g (x)} in f (y)", Set("x", "y"))
+    testCapture("fun (x) = {x + 1}", Set())
+    testCapture("let fac = fun (x) = {if (x) x*fac(x-1) else 1} in fac(5)", Set())
+    testCapture("let x=1 in let f=fun(y)={y+x} in f(1)", Set())
+    testCapture("let f=fun(y)={y+x} in f(1)", Set("x"))
+  }
+
+  @Test def runFunAndCall(): Unit = {
+    assertRun(2.5, "let x=1 in let f=fun(y)={y+x} in f(1.5)")
+    assertRun(2.5, "let f=fun(g)={g(1.0)} in let g=fun(x)={x+1.5} in f(g)")
```

Comment:
I see here that you somehow support higher-order functions.

I guess this works because your compiler basically treats the program as dynamically typed, and defers *all* type checks at run-time. Writing a type-checker for this language with HoFs will be insanely hard, compared to the expected difficulty of this assignment. It goes well beyond what is expected here. Essentially, I believe you'll need Hindley-Milner type inference, which is Haskell-like, and not at all Java-like nor Scala-like.